src/System.Management.Automation/engine/hostifaces/LocalPipeline.cs

```diff
@@ -1306,6 +1310,32 @@ internal class PipelineStopper
         /// Object used for synchronization.
         /// </summary>
         private object _syncRoot = new object();
+
+        internal void AwaitPipelineCriticalSection()
+        {
+            _criticalSectionSemaphore.Wait();
+            _criticalSectionSemaphore.Release();
+        }
+
+        internal void EnterCriticalSection()
+        {
+            Interlocked.Increment(ref _criticalSectionDepth);
+            _criticalSectionSemaphore.Wait(millisecondsTimeout: 0);
+        }
+
+        internal void ExitCriticalSection()
+        {
+            if (Interlocked.Decrement(ref _criticalSectionDepth) == 0)
+            {
+                _criticalSectionSemaphore.Release();
+            }
+        }
+
+        internal bool InCriticalSection
+        {
+            get => Interlocked.CompareExchange(ref _criticalSectionDepth, value: 0, comparand: 0) != 0;
+        }
```

Comment:
Yeah, a read can still occur either side of this, and that's fine. I think the main intent here is to avoid this particular read from occurring before and pending writes to the field (i.e., some thread moves to increment the depth counter, a different thread moves to check the value, and the processor optimizes the order such that the read occurs before the write, even though the write call may have happened before the read).

I will fully and happily admit to not being an expert on how the heck Interlocked works, but that's my understanding (and intent) from reading over the Interlocked documentation.

Reply:
I do agree that modifying the underlying counter isn't a _great_ thing to have in there and it would be nice if we can avoid it, but I don't know that there's a simple `Interlocked.Read()` API we can use there. I recall looking for one and coming up empty-handed.

Reply:
>  I think the main intent here is to avoid this particular read from occurring before any pending writes to the field (i.e., some thread moves to increment the depth counter, a different thread moves to check the value, and the processor optimizes the order such that the read occurs before the write, even though the write call may have happened before the read).

Yeah, my point is basically that `Interlocked` does nothing to help you with that here. The whole interlocked read operation is atomic, but a write could occur either side of it.

You're also not modifying the counter; because the `value` and `comparand` are the same, you'll always end up with the same value. It's basically saying "if x is 0, set x to 0".

So you're using a threading primitive here, but it's not helping you manage the thread state at all.

> but I don't know that there's a simple Interlocked.Read() API we can use there

Reads are already atomic. The `Interlocked` APIs wrap single native instructions that replace multiple sequential instructions; so `Interlocked.Increment()` replaces `load $register $addr; add $register 1; store $register $addr` with something like `atomic_add $addr 1`. It's the same as just adding an unshared lock inside the method, so your code is currently equivalent to this:

```csharp
internal bool InCriticalSection
{
    get
    {
        lock (_getterLock)
        {
            int depth = _criticalSectionDepth;
            if (_criticalSectionDepth == 0)
            {
                _criticalSectionDepth = 0;
            }
            return depth;
        }
    }
}
```

When I say *unshared* lock, I mean that the writer methods don't take out `_getterLock`. So this provides no protection from writers, and the `CompareExchange` logic doesn't do anything useful.